staging/src/k8s.io/apimachinery/pkg/util/net/http.go

```diff
@@ -189,6 +194,46 @@ func GetClientIP(req *http.Request) net.IP {
 	return net.ParseIP(req.RemoteAddr)
 }
 
+// Prepares the X-Forwarded-For header for another forwarding hop by appending the previous senders
+// IP address to the X-Forwarded-For chain.
+func PrepareForwardedFor(req *http.Request) {
+	remoteIP := GetRemoteIP(req)
+
+	hdr := req.Header
+	// First check the X-Forwarded-For header for requests via proxy.
+	hdrForwardedFor := hdr.Get("X-Forwarded-For")
+	if hdrForwardedFor != "" {
+		// Append the remote IP if necessary.
+		parts := strings.Split(hdrForwardedFor, ",")
+		for i := len(parts) - 1; i >= 0; i-- {
+			// Find the last non-empty part.
+			if parts[i] == "" {
+				continue
+			}
+			if ip := net.ParseIP(strings.TrimSpace(parts[i])); ip.Equal(remoteIP) {
+				// Remote IP already appended; nothing to do.
+				return
+			}
+			parts = append(parts[:i+1], remoteIP.String())
+			hdr.Set("X-Forwarded-For", strings.Join(parts, ","))
+			return
+		}
+		// Empty header, handle as if the header was not set.
+		hdrForwardedFor = ""
+	}
+
+	// Try the X-Real-Ip header.
+	hdrRealIp := hdr.Get("X-Real-Ip")
```

Comment:
> how is it determined and trusted?

Some proxies set it. I don't think it's very well defined... It can't be trusted. That's why we still append the `RemoteAddr` IP below.

> we only insert this into X-Forwarded-For when there is no existing X-Forwarded-For header?

Right. See above comment.

> which should take priority, the remote IP obtained from GetRemoteIP() or this header?

The logic in `PrepareForwardedFor` is based off the existing logic in [`GetClientIP`](https://github.com/kubernetes/kubernetes/blob/effca9605bdac66421c249f21bc3b98bc735dcfe/staging/src/k8s.io/apimachinery/pkg/util/net/http.go#L156), which gives preference to the `X-Real-IP` header.  Note that for audit logging, we're going to log the full `X-Forwarded-For` chain, since there's nothing stopping a client from adding bogus IPs to the front of the chain, or spoofing `X-Real-IP`.